h2/src/main/org/h2/mvstore/MVMap.java

```diff
@@ -1241,59 +1294,422 @@ public String toString() {
+                    pos = pos.parent;
+                }
+                if(needUnlock) {
+                    unlockRoot(p, attempt);
+                    needUnlock = false;
+                } else if(!updateRoot(rootReference, p, attempt)) {
+                    decisionMaker.reset();
+                    continue;
+                }
+                while (tip != null) {
+                    tip.page.removePage();
+                    tip = tip.parent;
+                }
+                if (store.getFileStore() != null) {
+                    store.registerUnsavedPage(unsavedMemory);
+                }
+                return result;
+            } finally {
+                if(needUnlock) {
+                    unlockRoot(rootReference.root, attempt);
+                }
+            }
+        }
+    }
+
+    private boolean lockRoot(DecisionMaker<? super V> decisionMaker, RootReference rootReference,
+                             int attempt, int contention) {
+        boolean success = lockRoot(rootReference);
+        if (!success) {
+            decisionMaker.reset();
+            if(attempt > 4) {
+                if (attempt <= 24) {
+                    Thread.yield();
+                } else {
+                    try {
+                        Thread.sleep(0, 100 / contention + 50);
+                    } catch (InterruptedException ignore) {/**/}
+                }
+            }
+        }
+        return success;
+    }
+
+    private boolean lockRoot(RootReference rootReference) {
+        return !rootReference.semaphore
+            && root.compareAndSet(rootReference, new RootReference(rootReference));
+    }
+
+    private void unlockRoot(Page newRoot, int attempt) {
+        boolean success;
+        do {
+            RootReference rootReference = getRoot();
+            RootReference updatedRootReference = new RootReference(rootReference, newRoot, attempt);
+            success = root.compareAndSet(rootReference, updatedRootReference);
+        } while(!success);
+    }
+
+    public static CursorPos traverseDown(Page p, Object key) {
+        CursorPos pos = null;
+        while (!p.isLeaf()) {
+            assert p.getKeyCount() > 0;
+            int index = p.binarySearch(key) + 1;
+            if (index < 0) {
+                index = -index;
+            }
+            pos = new CursorPos(p, index, pos);
+            p = p.getChildPage(index);
+        }
+        return new CursorPos(p, p.binarySearch(key), pos);
+    }
+
+    private static final class EqualsDecisionMaker<V> extends DecisionMaker<V> {
+        private final DataType dataType;
+        private final V        expectedValue;
+        private       Decision decision;
+
+        private EqualsDecisionMaker(DataType dataType, V expectedValue) {
+            this.dataType = dataType;
+            this.expectedValue = expectedValue;
+        }
+
+        @Override
+        public Decision decide(V existingValue, V providedValue) {
```

Comment:
I don't think I understand the question. It does use providedValue to select between PUT and DELETE.

Reply:
Ok, now I got it! selectValue() does not use existingValue parameter, indeed. That's because whole thing was back-ported, and subsequent PR for TransactionStore and friends will have subclasses with use this parameter.